### ISSUE TITLE
Create dependencies concurrently 

### ISSUE THREAD

**Issue by Tishka17** [reagento/dishka#45]

**Comment by FRidh** [comment]
As I have a need for this functionality I gave it a go. Please see the 2 PR's with different solutions to the problem. https://github.com/reagento/dishka/pull/704 https://github.com/reagento/dishka/pull/705

**Comment by Tishka17** [comment]
@FRidh can you please share your UseCase?

**Comment by Tishka17** [comment]
My thought on concurrent creation:

1. Both sync and async should be implemented. Though, it can be done in separate PR's we need to clarify API design here
3. We have generators and factories, both can be evaluated concurrently
4. It should handle errors and cancellation of parent task, so no work can continue if any factory failed and `container.get` raised an error. 
5. No factory should be called twice because of concurrent execution (diamond problem, noted in PRs)
6. Tests should check behavior, not timing. Synchronization primitives can be used for that
7. Regardles of the result (success or error), no background tasks or intermediate objects should be left after `container.get` finished.
2. Concurrent execution must be opt-in. It affects performance and safety, so it is not always safe
8. It is open question if we want this be enabled per-factory or once on container creation. E.g "normal" factories are called directly, but thoses, marked as "heavy" executed in background.
9. It is open question if we want concurrent finalization
10. Effect on performace should be discussed
11. Performance impact should be discussed
12. We should support `trio` without using `anyio` as it affects performance

**Comment by FRidh** [comment]
Thanks for the prompt feedback.

> [@FRidh](https://github.com/FRidh) can you please share your UseCase?

We have a web application with a compute backend that does various numerical calculations that we need to orchestrate. These form a rather complex DAG so in order to manage the complexity I want to use inversion of control and in this case inject not just services (which we currently do) but instead primarily artifacts. Think of it like orchestration tools such as Dagster but then used as part of a user application. I imagine this is not how this library is commonly used.

For this use case I am interested in async evaluation of the artifacts. Anything that is compute heavy is offloaded to workers or other third-party services. We don't need to support a high number of concurrent tasks so performance is less important but separating the wiring from the business logic absolutely is. Not supporting concurrent evaluation is however a blocker and hence I gave Claude a quick go to evaluate the feasibility and get initial feedback, especially as I am not yet familiar with the dishka internals.


> My thought on concurrent creation:
> 
>     1. Both sync and async should be implemented. Though, it can be done in separate PR's we need to clarify API design here

What do you have in mind for the sync container. Let the user provide a pool executor? What if the user generates a pool and uses a pool inside of a dependency? I suppose only document that this can cause trouble but trying to handle this in any way is out of scope. 

I could imagine a `concurrency` parameter that accepts a pool or semaphore.

>     2. We have generators and factories, both can be evaluated concurrently

>     3. It should handle errors and cancellation of parent task, so no work can continue if any factory failed and `container.get` raised an error.
 
This would get easier with TaskGroup over gather. I noticed the pyproject.toml said Python 3.10+ is supported. Is this a requirement? Reading up on trio I suppose it makes sense to use it for this.

>     4. No factory should be called twice because of concurrent execution (diamond problem, noted in PRs)
> 
>     5. Tests should check behavior, not timing. Synchronization primitives can be used for that
> 
>     6. Regardles of the result (success or error), no background tasks or intermediate objects should be left after `container.get` finished.
> 
>     7. Concurrent execution must be opt-in. It affects performance and safety, so it is not always safe

I suggest here the `concurrency` parameter.

>     8. It is open question if we want this be enabled per-factory or once on container creation. E.g "normal" factories are called directly, but thoses, marked as "heavy" executed in background.
 
Supporting per-factory is adding complexity. Perhaps that could be a future improvement?

>     9. It is open question if we want concurrent finalization

> 
>     10. Effect on performace should be discussed
> 
>     11. Performance impact should be discussed
> 
>     12. We should support `trio` without using `anyio` as it affects performance

Do you mean supporting trio as event loop for calling say container.get() or also supporting trio code as a @provide dependency?

I haven't used trio yet myself but will look into it. I'll get back regarding the open items.

**Comment by Tishka17** [comment]
Currently we do not use loop anyhow, so the code just works with trio. But if we switch to TaskGroup we need to have explicit support. I do not think it is a problem to support new feature only for python 3.11+

Should we provide some "Executor" primitive to manage concurrent dependencies? I guess we can allow user to customize behavior of code builder here with fallback to current behavior. Probably, we could cover all cases having multiple implementations of such class

**Comment by FRidh** [comment]
I could imagine yes the user provides their concurrency strategy and dishka would provide several out of the box. 

For managing requirements I like to use spec-kit these days, see the [requirements](https://github.com/FRidh/dishka/blob/001-concurrent-dependency-creation/specs/001-concurrent-dependency-creation/spec.md) in [branch](https://github.com/FRidh/dishka/tree/001-concurrent-dependency-creation) (see also fork [PR](https://github.com/FRidh/dishka/pull/1)).

> Should we provide some "Executor" primitive to manage concurrent dependencies? I guess we can allow user to customize behavior of code builder here with fallback to current behavior. Probably, we could cover all cases having multiple implementations of such class

This seems doable. I think we can also add support incrementally:
1. basic async asyncio strategy
2. basic sync strategy
3. async trio strategy
4. add codegen support
5. per-factory dispatching

All while not breaking any interfaces. Note per-factory dispatching does need further investigation. I think this interface could work but I am not sure yet about this.

Following is a generated summary of the [interfaces](https://github.com/FRidh/dishka/blob/001-concurrent-dependency-creation/specs/001-concurrent-dependency-creation/contracts/concurrency_api.md):                                                                                            

---
<details>

## Public Interface: Concurrent Dependency Creation

### User-Facing API

The only thing users interact with — pass a strategy to the container:

```python
make_async_container(*providers, concurrency=AsyncioStrategy())
make_container(*providers, concurrency=ThreadPoolStrategy())
```

Omitting `concurrency=` keeps current sequential behavior with zero overhead.

### Built-in Strategies

Pick one and pass it to the container. That's it.

| Strategy | Use case |
|----------|----------|
| `AsyncioStrategy()` | Async container, unlimited concurrency |
| `AsyncioSemaphoreStrategy(max_concurrent=N)` | Async container, bounded concurrency |
| `TrioStrategy()` | Async container under trio |
| `ThreadPoolStrategy(executor=None)` | Sync container, thread-based parallelism |
| `ProcessPoolStrategy(executor=None)` | Sync container, process-based parallelism |

### Strategy Protocols (for custom strategy authors)

Only relevant if you're writing your own strategy. The resolver calls `run()` internally with independent callables from each topological layer — users never call this directly.

```python
class AsyncConcurrencyStrategy(Protocol):
    async def run(
        self,
        factories: Sequence[tuple[DependencyKey, Callable[[], Awaitable[object]]]],
    ) -> Sequence[object]: ...

class SyncConcurrencyStrategy(Protocol):
    def run(
        self,
        factories: Sequence[tuple[DependencyKey, Callable[[], object]]],
    ) -> Sequence[object]: ...
```

`DependencyKey` is included alongside each callable so custom strategies can route per-dependency if needed. Built-in strategies ignore it.

### Future Extension

An optional `compile()` method can be added to the protocols for codegen optimization — no breaking changes to the above.

</details>

**Comment by Tishka17** [comment]
>  per-factory dispatching does need further investigation. 

I agree this might be really tough to implement, though I need to have ideas on API

> 4. add codegen support

I suggest doing this at first as it affects API of strategy.

**Comment by FRidh** [comment]
I explored this further, see https://github.com/FRidh/dishka/pull/1 on my "fork". Please have a look at the docs 

What I don't like yet is the per factory dispatching. I had initially in mind being able to specify multiple executors, but then having a strategy made sense as well, and leaving it to the strategy to choose whether to use multiple executors or not.

On this PR there is support for per-factory "tags". The code generating `.compile` does not yet support this, I'll look into that further. Note the current `.compile` interface is the original one I had in mind for `.compile` and `.run`.

**Comment by FRidh** [comment]
Regarding the changes, can we include in this repo
- the dependency group changes in `pyproject.toml` as this is part of a PEP. What is your opinion on including a uv lock file?
- benchmarks

I could deliver these in separate PR's.

**Comment by Tishka17** [comment]
1. Not clear for me, what do you suggest doing with dependency groups?
2. I do not see any usecase for uv lock file for the library.
3. We have a separate issue for benchmarks, but I am very unsure what to we want from them.

**Comment by FRidh** [comment]
It makes it easier to run the tests and other tasks.

A lock file has no use for a library though it does allow for reproducibility for test executation.

I created benchmarks to show the performance difference between different strategies. This can help detect regressions.

**Comment by FRidh** [comment]
Regarding the PR I'd like feedback on whether the approach and interfaces are acceptable. Do we want this approach with the strategy protocol? At this stage the actual code should not matter yet.

The test coverage demonstrates the topics brought up have been implemented and work showing all parts *can* be done.

**Comment by Tishka17** [comment]
I was thinking about locking dependencies for tests, but it will 1 additional lock file per dependencies file here:  https://github.com/reagento/dishka/tree/develop/requirements
It is still unclear for me how to do minimal updates in lock files. Afaik they are usually just regenerated with latest version which is not what I expect once we introduce them)

**Comment by Tishka17** [comment]
> Regarding the PR I'd like feedback on whether the approach and interfaces are acceptable. Do we want this approach with the strategy protocol? At this stage the actual code should not matter yet.

Sorry, I do not see how we can solve noted requirements iwth such protocol. Anyway I see usefuly only the strategy that can be used in code generation
